File: sdk/cpp/src/signing.cpp
#include "cap/signing.h"
#include "cap/codec.h"

#include <cstring>
#include <string>
#include <vector>

#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/pem.h>

namespace cap {

bool SignPacket(cordum::agent::v1::BusPacket* packet, EVP_PKEY* private_key) {
  if (!packet || !private_key) return false;

  auto unsigned_bytes = MarshalUnsignedForSignature(*packet);
  if (unsigned_bytes.empty() && packet->ByteSizeLong() > 0) return false;

  // SHA-256 hash
  unsigned char hash[EVP_MAX_MD_SIZE];
  unsigned int hash_len = 0;
  if (EVP_Digest(unsigned_bytes.data(), unsigned_bytes.size(), hash, &hash_len,
                 EVP_sha256(), nullptr) != 1) {
    return false;
  }

  // ECDSA sign the hash
  EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(private_key, nullptr);
  if (!ctx) return false;

  bool ok = false;
  if (EVP_PKEY_sign_init(ctx) == 1) {
    size_t sig_len = 0;
    if (EVP_PKEY_sign(ctx, nullptr, &sig_len, hash, hash_len) == 1) {
      std::vector<unsigned char> sig(sig_len);
      if (EVP_PKEY_sign(ctx, sig.data(), &sig_len, hash, hash_len) == 1) {
        packet->set_signature(sig.data(), sig_len);
        ok = true;
      }
    }
  }

  EVP_PKEY_CTX_free(ctx);
  return ok;
}

bool VerifyPacketSignature(const cordum::agent::v1::BusPacket& packet,
                           EVP_PKEY* public_key) {
  if (!public_key) return false;

  const std::string& sig = packet.signature();
  if (sig.empty()) return false;

  auto unsigned_bytes = MarshalUnsignedForSignature(packet);

  // SHA-256 hash
  unsigned char hash[EVP_MAX_MD_SIZE];
  unsigned int hash_len = 0;
  if (EVP_Digest(unsigned_bytes.data(), unsigned_bytes.size(), hash, &hash_len,
                 EVP_sha256(), nullptr) != 1) {
    return false;
  }

  // ECDSA verify
  EVP_PKEY_CTX* ctx = EVP_PKEY_CTX_new(public_key, nullptr);
  if (!ctx) return false;

  bool ok = false;
  if (EVP_PKEY_verify_init(ctx) == 1) {
    ok = EVP_PKEY_verify(ctx, reinterpret_cast<const unsigned char*>(sig.data()),
                         sig.size(), hash, hash_len) == 1;
  }

  EVP_PKEY_CTX_free(ctx);
  return ok;
}

EVP_PKEY* LoadPrivateKey(const std::string& pem) {
  BIO* bio = BIO_new_mem_buf(pem.data(), static_cast<int>(pem.size()));
  if (!bio) return nullptr;

  EVP_PKEY* key = PEM_read_bio_PrivateKey(bio, nullptr, nullptr, nullptr);
  BIO_free(bio);
  return key;
}

EVP_PKEY* LoadPublicKey(const std::string& pem) {
  BIO* bio = BIO_new_mem_buf(pem.data(), static_cast<int>(pem.size()));
  if (!bio) return nullptr;

  EVP_PKEY* key = PEM_read_bio_PUBKEY(bio, nullptr, nullptr, nullptr);
  BIO_free(bio);
  return key;
}

}  // namespace cap

File: sdk/go/signing.go
package capsdk

import (
	"crypto/ecdsa"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"

	agentv1 "github.com/cordum-io/cap/v2/cordum/agent/v1"
	"google.golang.org/protobuf/proto"
)

var (
	// ErrMissingSignature indicates a packet did not include a signature.
	ErrMissingSignature = errors.New("capsdk: missing signature")
	// ErrInvalidSignature indicates a signature could not be verified.
	ErrInvalidSignature = errors.New("capsdk: invalid signature")
)

// MarshalDeterministic serializes a protobuf message with deterministic ordering.
func MarshalDeterministic(msg proto.Message) ([]byte, error) {
	if msg == nil {
		return nil, errors.New("capsdk: message is nil")
	}
	return proto.MarshalOptions{Deterministic: true}.Marshal(msg)
}

// MarshalUnsignedForSignature serializes a BusPacket with signature cleared
// using CAP's cross-SDK signing order.
func MarshalUnsignedForSignature(packet *agentv1.BusPacket) ([]byte, error) {
	if packet == nil {
		return nil, errors.New("capsdk: packet is nil")
	}
	clone := proto.Clone(packet).(*agentv1.BusPacket)
	clone.Signature = nil
	return marshalPacketForSignature(clone)
}

// SignPacket signs a BusPacket and stores the signature on the packet.
func SignPacket(packet *agentv1.BusPacket, key *ecdsa.PrivateKey) error {
	if packet == nil {
		return errors.New("capsdk: packet is nil")
	}
	if key == nil {
		return errors.New("capsdk: private key is nil")
	}

	unsignedData, err := MarshalUnsignedForSignature(packet)
	if err != nil {
		return fmt.Errorf("marshal unsigned packet: %w", err)
	}

	hash := sha256.Sum256(unsignedData)
	signature, err := ecdsa.SignASN1(rand.Reader, key, hash[:])
	if err != nil {
		return fmt.Errorf("sign packet: %w", err)
	}
	packet.Signature = signature
	return nil
}

// VerifyPacketSignature verifies a BusPacket signature using the provided public key.
func VerifyPacketSignature(packet *agentv1.BusPacket, key *ecdsa.PublicKey) error {
	if packet == nil {
		return errors.New("capsdk: packet is nil")
	}
	if key == nil {
		return errors.New("capsdk: public key is nil")
	}
	signature := packet.GetSignature()
	if len(signature) == 0 {
		return ErrMissingSignature
	}

	unsignedData, err := MarshalUnsignedForSignature(packet)
	if err != nil {
		return fmt.Errorf("marshal unsigned packet: %w", err)
	}
	hash := sha256.Sum256(unsignedData)
	if !ecdsa.VerifyASN1(key, hash[:], signature) {
		return ErrInvalidSignature
	}
	return nil
}

func marshalPacketForSignature(packet *agentv1.BusPacket) ([]byte, error) {
	if packet.GetAuthToken() == "" || packet.GetPayload() == nil {
		return MarshalDeterministic(packet)
	}
	// Python protobuf and protobufjs encode BusPacket oneof payload fields
	// before auth_token when signature is cleared. Go's generated fast path
	// emits scalar fields before oneofs, so auth_token would otherwise be
	// signed before the payload and fail cross-SDK verification.
	clone := proto.Clone(packet).(*agentv1.BusPacket)
	authToken := clone.AuthToken
	clone.AuthToken = ""

	data, err := MarshalDeterministic(clone)
	if err != nil {
		return nil, err
	}
	authBytes, err := MarshalDeterministic(&agentv1.BusPacket{AuthToken: authToken})
	if err != nil {
		return nil, err
	}
	return append(data, authBytes...), nil
}
